ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala

```diff
@@ -44,14 +46,28 @@ class KeyValueSubmission(metrics: Metrics) {
       tx: SubmittedTransaction,
   ): DamlSubmission =
     metrics.daml.kvutils.submission.conversion.transactionToSubmission.time { () =>
-      val inputDamlStateFromTx = InputsAndEffects.computeInputs(tx, meta)
       val encodedSubInfo = buildSubmitterInfo(submitterInfo)
+      val packageIdStates = meta.optUsedPackages
+        .getOrElse(
+          throw new InternalError("Transaction was not annotated with used packages")
+        )
+        .map(Conversions.packageStateKey)
+      val partyStates =
+        (tx.informees ++ submitterInfo.actAs).toList.map(Conversions.partyStateKey)
+      val contractIdStates =
+        tx.inputContracts[ContractId].map(Conversions.contractIdToStateKey)
+      val contractKeyStates =
```

Comment:
Why do we treat contract keys created with the transaction as inputs but don’t do the same for contract ids? That seems a bit inconsistent. I assume it has something to do with duplicate key detection or something along those lines?

Reply:
As far I understand, for validation, KV have to collect all the info needed to replay the transaction. For created contract it should know the contract key does not exists.

Reply:
alright so the input is the “non-existence of the key” or something like that, makes sense :+1: